APMrover2/mode_follow.cpp

```diff
@@ -0,0 +1,82 @@
+#include "mode.h"
+#include "Rover.h"
+
+// initialize follow mode
+bool ModeFollow::_enter()
+{
+    // re-use guided mode
+    gcs().send_text(MAV_SEVERITY_WARNING, "Follow-mode init");
+    if (!g2.follow.enabled()) {
+        return false;
+    }
+    return ModeGuided::enter();
+}
+
+void ModeFollow::update()
+{
+    // variables to be sent to velocity controller
+    Vector3f desired_velocity_neu_cms;
+    float yaw_cd = 0.0f;
+
+    Vector3f dist_vec;  //vector to lead vehicle
+    Vector3f dist_vec_offs; // vector to lead vehicle + offset
+    Vector3f vel_of_target; // velocity of lead vehicle
+
+    if (g2.follow.get_target_dist_and_vel_ned(dist_vec, dist_vec_offs, vel_of_target)) {
+
+        // convert dist_vec_offs to cm in NEU
+        const Vector3f dist_vec_offs_neu(dist_vec_offs.x * 100.f, dist_vec_offs.y * 100.0f, -dist_vec_offs.z * 100.0f);
+
+        // calculate desired velocity vector in cm/s in NEU
+        const float kp = g2.follow.get_pos_p().kP();
+        desired_velocity_neu_cms.x = (vel_of_target.x * 100.0f) + (dist_vec_offs_neu.x * kp);
+        desired_velocity_neu_cms.y = (vel_of_target.y * 100.0f) + (dist_vec_offs_neu.y * kp);
+
+        // scaled desired velocity to stay within horizontal speed limit
+        float desired_speed_xy = safe_sqrt(sq(desired_velocity_neu_cms.x) + sq(desired_velocity_neu_cms.y));
+        if (!is_zero(desired_speed_xy) && (desired_speed_xy > pos_control->get_speed_xy())) {
+            const float scalar_xy = pos_control->get_speed_xy() / desired_speed_xy;
+            desired_velocity_neu_cms.x *= scalar_xy;
+            desired_velocity_neu_cms.y *= scalar_xy;
+            desired_speed_xy = pos_control->get_speed_xy();
+        }
+
+        // unit vector towards target position (i.e. vector to lead vehicle + offset)
+        Vector3f dir_to_target_neu = dist_vec_offs_neu;
+        const float dir_to_target_neu_len = dir_to_target_neu.length();
+        if (!is_zero(dir_to_target_neu_len)) {
+            dir_to_target_neu /= dir_to_target_neu_len;
+        }
+
+        // create horizontal desired velocity vector (required for slow down calculations)
+        Vector2f desired_velocity_xy_cms(desired_velocity_neu_cms.x, desired_velocity_neu_cms.y);
+
+        // create horizontal unit vector towards target (required for slow down calculation)
+        Vector2f dir_to_target_xy(desired_velocity_xy_cms.x, desired_velocity_xy_cms.y);
+        if (!dir_to_target_xy.is_zero()) {
+            dir_to_target_xy.normalize();
+        }
+
+        // slow down horizontally as we approach the target (use 1/2 of maximum deceleration for gentle slow down)
+        const float dist_to_target_xy = Vector2f(dist_vec_offs_neu.x, dist_vec_offs_neu.y).length();
+        g2.avoid.limit_velocity(pos_control->get_pos_xy_p().kP().get(), pos_control->get_accel_xy() * 0.5f, desired_velocity_xy_cms, dir_to_target_xy, dist_to_target_xy, rover.G_Dt);
```

Comment:
these calls to the pos_control should be removed/replaced with the equivalents for rover.  mode_loiter.cpp may have some parts that can be borrowed to replace what's here.  Also mode.cpp's calc_throttle includes the calls to g2.avoid.adjust_speed which should be used instead of adjust_velocity()... or maybe the call to avoidance here simply isn't required anymore because it'll be done within calc_throttle().